### ISSUE TITLE
브랜치 전략: feature-baseline-training

### ISSUE THREAD

**Issue by iejob** [iejob/computer-vision-competition-1SEN#9]
## 목표와 범위
- **목표:**
    기본 학습 파이프라인을 표준화하여 **재현성·안정성·가독성**을 갖춘 베이스라인 모델을 구축합니다.
    
    - 본 브랜치는 **백본 선택(ResNet/EfficientNet/ConvNeXt/소형 ViT), 손실·옵티마이저·스케줄러·AMP/EMA·체크포인트·조기종료**를 포함한 **엔드투엔드 훈련 루틴**을 제공합니다.
    - 모든 기능은 **`ipynb`에서 사전 검증** 후, 통과 코드만 `src/`의 표준 디렉터리로 **모듈화 이관**합니다.
    - 병합 전 **훈련 성공률·재현성·로그/아티팩트 완결성·Macro-F1 기준 충족**을 DoD로 정의합니다.
- **범위:**
    
    단일/다중 폴드 학습 루프, 손실/가중치/라벨 스무딩, 옵티마이저(AdamW 기본), 스케줄러(Cosine+Warmup 기본), AMP(fp16/bf16), EMA, Gradient Clipping/Accumulation, EarlyStopping, 체크포인트/로그·플롯 산출, `feature-data-eda-loader`/`feature-eval-cv` 연계.
    
    - **데이터·증강**은 `feature-data-eda-loader` 산출물 사용.
    - **평가·리포팅**은 `feature-eval-cv`에 위임하되, 학습 중 **검증 Macro-F1**을 실시간 기록/저장.
    - **재현성 가드**: 설정/분할 **sha256 해시 잠금**, **라벨 정렬(0..16) 고정**, **Resume/Failover** 내장.

---

### 1) 구성·설정(단일 소스 & 해시 잠금)
- **설정 파일:** `configs/train_config.json`
    - 백본/해상도/배치/에폭/옵티마/스케줄러/손실/라벨스무딩/클리핑/EMA/AMP/정밀도/체크포인트/로그 경로/eval 주기/재개 옵션/로더 내성 등.
- **런 아이디 규약:** `YYYYMMDD_HHMMSS_<arch>_img<size>_bs<batch>` → `experiments/train/<run_id>/` 생성.
- **해시 잠금:** 실행 시 `train_config.json`과(사용 시) `data/processed/folds_k5_seed42.csv`의 **sha256**을 산출·저장(`config.hash`, `folds.hash`)하고 재개 시 **불일치면 즉시 실패**.

### 2) 모델·손실·최적화(Param-Group 표준)
- **백본(사전학습 허용):** `resnet50`, `efficientnet_b3`, `convnext_tiny`, `vit_base_patch16_224(선택)`
- **헤드:** `nn.Linear(in_features, 17)` + 선택 Dropout(0.2~0.5)
- **손실:** CrossEntropy(+ `class_weights.json` 옵션), Label Smoothing(ε=0.1), Focal(γ=2 옵션)
- **옵티마이저:** AdamW(lr=3e-4, wd=1e-4 기본)
    - **Param groups:** `bias/Norm` 계열 **no_decay(=0)**, 그 외 `weight_decay` 적용
- **스케줄러:** CosineAnnealingLR(T_max=epochs) + Linear Warmup(5%), 옵션 OneCycleLR
- **정규화·안정화:** Gradient Clipping(‖g‖≤1.0), **AMP(fp16/bf16)** + GradScaler, **EMA(decay=0.999)**(검증·저장 시 EMA 가중치 사용 옵션)

### 3) 학습 루프·콜백(Resume/Failover 내장)
- **루프:** train/val 분리, epoch 종료 시 **라벨 정렬 고정(0..16)**로 Macro-F1 산출(`src/utils/metrics.py`)
- **콜백:**
    - `ModelCheckpoint(top_k=2, monitor='macro_f1', mode='max', save_every_n_epochs)`
    - `EarlyStopping(patience=5, delta=0.0005, monitor='macro_f1')`
    - `LRSchedulerStep(per_epoch/per_step)`
    - `EMAUpdate(per_step)`
    - `HistoryLogger`(CSV/JSON: lr, loss, macro_f1, **imgs/sec**, grad_norm)
- **실패 정책:** NaN/Inf loss 감지 시 즉시 중단 → **직전 안정 체크포인트로 1회 롤백 재시도**(옵션).
- **체크포인트 스키마(표준):**
    
    ```python
    {
      "model", "optimizer", "scheduler", "scaler", "ema",
      "epoch", "global_step", "best_macro_f1",
      "rng_state": {"torch","cuda","numpy","python"},
      "config_hash", "folds_hash"
    }
    ```
    
- **Resume:** `resume_from` + `strict_resume=true` 시 **옵티마/스케줄러/EMA/스케일러/epoch/시드** 완전 복원.

### 4) 데이터 연계
- `DocTypeDataset`, `get_train_transforms()`, `get_val_transforms()` 사용.
- 샘플링: **WeightedRandomSampler**(선택) 또는 균등.
- `data/processed/class_weights.json` 존재 시 손실 가중치 적용.
- 로더 내성: `max_loader_errors`(기본 0), **연속 오류 시 중단**, 워커 시드 고정, `persistent_workers=True`.

### 5) 단일 폴드/다중 폴드 학습(+ OOF 훅 옵션)
- **단일 폴드:** `train/val` CSV로 학습 → `best.pth` 저장.
- **다중 폴드(K=5):** `folds_k5_seed42.csv` 기반, fold별 독립 `run_id`.
    - **옵션 OOF 훅:** 각 fold 검증 단계에서 `val_probs.npy`, `val_targets.npy`, `val_ids.csv` 저장(기본 off).

### 6) 정밀도·성능
- **precision:** `"fp16"|"bf16"|"fp32"`(기본 `"fp16"`). **평가·집계는 fp32 강제**.
- `torch.set_float32_matmul_precision("high")`(선택).
- **스루풋 로그:** 워밍업 제외 **imgs/sec**를 `throughput.txt`에 저장.

### 7) 로깅·아티팩트
- **실시간 로그:** `train_log.csv`(epoch, lr, loss, macro_f1, time, imgs/sec)
- **곡선 플롯:** `loss_curve.png`, `macro_f1_curve.png`, `lr_curve.png`
- **체크포인트:** `best.pth`, `last.pth`, `topk/epoch=.._macro=..pth`
- **메타:** `config.json`, `env.json`(패키지/디바이스), `summary.json`(최종 성과), **`config.hash`, `folds.hash`**
- (KFold) `fold_<k>/*` + (옵션) `val_*` 3종

### 8) CLI/CI 스모크
- **CLI:**
    - `python -m src.training.train --cfg configs/train_config.json --run_id <tag>`
    - `python -m src.training.train --cfg configs/train_config.json --run_id <tag> --resume_from experiments/train/<run_id>/last.pth`
- **CI 스모크:**
    - `pytest -q tests/test_train_minibatch.py`(1배치 학습),
    - `python -m src.training.train --cfg configs/train_config.json --run_id ci_smoke --fast_dev_run`

---

## ipynb-우선 워크플로우 (파일명·셀 단위 체크리스트)
1. **`notebooks/TRAIN/00_train_config_and_env.ipynb`**
    - [ ]  `configs/train_config.json` 작성/검증(경로·하이퍼)
    - [ ]  환경 로그 수집 → `env.json` 저장, **`config.hash`/`folds.hash` 생성**
2. **`notebooks/TRAIN/01_single_fold_baseline_train.ipynb`**
    - [ ]  ResNet50, 224, bs=64, ep=20 단일 폴드 학습
    - [ ]  `best.pth`, `train_log.csv`, 곡선 플롯 저장
3. **`notebooks/TRAIN/02_kfold_training_runner.ipynb`**
    - [ ]  K=5 순회 학습, fold별 `best.pth`/검증 지표 저장
    - [ ]  (옵션) OOF 훅: `val_probs/val_targets/val_ids` 저장
4. **`notebooks/TRAIN/03_ablation_loss_aug_opt.ipynb`**
    - [ ]  CE vs CE+LS vs Focal, AdamW vs SGD, Sampler on/off 비교
5. **`notebooks/TRAIN/04_training_plots_and_checkpoints.ipynb`**
    - [ ]  로그/곡선/체크포인트 검증(누락·손상 0)
    - [ ]  best epoch 타당성 점검
6. **`notebooks/TRAIN/05_export_training_modules.ipynb`**
    - [ ]  통과 코드만 `src/training/*.py`, `src/models/*.py`, `src/utils/*.py` 이관
7. **`notebooks/TRAIN/06_resume_and_fault_tolerance.ipynb`** *(보강)*
    - [ ]  3epoch 후 중단→재개: 최종 Macro-F1 **±0.1%p** 이내
    - [ ]  NaN/Inf 롤백 재시도 동작 확인
8. **`notebooks/TRAIN/07_cli_and_ci_smoke.ipynb`** *(선택)*
    - [ ]  CLI 실행 스모크, `pytest` 최소 스위트 통과

---

## 산출물(Artifacts)
- **코드 모듈**
    - `src/models/backbone.py` : `build_model(arch, num_classes, drop, pretrained=True)`
    - `src/training/train.py` : `train_one_epoch`, `validate_one_epoch`, `fit(...)`(resume/failover/ema 지원)
    - `src/training/callbacks.py` : `ModelCheckpoint`, `EarlyStopping`, `EMA`, `HistoryLogger`
    - `src/training/optim.py` : AdamW param-groups( no_decay for bias/Norm ), 스케줄러 팩토리
    - `src/training/utils.py` : AMP/GradScaler, grad clipping, metrics hook, worker_init
    - `src/utils/reproducibility.py` : 시드 고정, CUDNN 설정(deterministic/benchmark 플래그)
- **리포트/로그**
    - `experiments/train/<run_id>/{config.json, config.hash, folds.hash, env.json, summary.json, train_log.csv, throughput.txt, loss_curve.png, macro_f1_curve.png, lr_curve.png, best.pth, last.pth, topk/...}`
    - (KFold) `experiments/train/<run_id>/fold_<k>/*` + (옵션) `val_probs.npy/val_targets.npy/val_ids.csv`
- **설정 파일**
    - `configs/train_config.json`
- **노트북(최종 파일명 확정)**
    - `notebooks/TRAIN/00_train_config_and_env.ipynb`
    - `notebooks/TRAIN/01_single_fold_baseline_train.ipynb`
    - `notebooks/TRAIN/02_kfold_training_runner.ipynb`
    - `notebooks/TRAIN/03_ablation_loss_aug_opt.ipynb`
    - `notebooks/TRAIN/04_training_plots_and_checkpoints.ipynb`
    - `notebooks/TRAIN/05_export_training_modules.ipynb`
    - `notebooks/TRAIN/06_resume_and_fault_tolerance.ipynb`
    - `notebooks/TRAIN/07_cli_and_ci_smoke.ipynb`

---

### 1 학습 파이프라인
- **입력:** `train.csv`, (선택) `folds_k5_seed42.csv`, 이미지 루트, 변환/샘플러.
- **출력:** 체크포인트, 로그/플롯, 요약 JSON, (KFold) fold별 결과, (옵션) OOF val 예측.
- **라벨 정렬 고정:** 모든 지표 계산 시 `[0..16]` 고정(누락 클래스 0 채움).
- **에러 내성:** 배치 로딩 오류 **연속 `max_loader_errors` 초과 시 중단**.

### 2 모델·옵티마이저·스케줄러
- timm 사전학습 백본 + 헤드 교체.
- AdamW(Param-Group no_decay), Cosine+Warmup(5%), OneCycle 옵션.

### 3 손실·가중치·정규화
- CE(+class_weights), Label Smoothing(0.05~~0.2), Focal(γ 1~~2).
- global-norm clip ≤ 1.0. EMA(decay 0.999, ep≥3부터).

### 4 정밀도·성능
- AMP(fp16/bf16), **평가·집계 fp32 강제**, imgs/sec 로깅.

### 5 검증·지표
- epoch마다 val 평가(또는 `eval_interval`).
- 주지표 Macro-F1, 보조 loss/accuracy. Best는 Macro-F1 기준.

### 6 체크포인트·재개
- **표준 스키마** 필수 키 포함, `strict_resume` 시 완전 복원.
- NaN/Inf 감지 시 롤백 재시도 1회(옵션).

### 7 재현성·감사
- 시드=42 고정, `config.hash`/`folds.hash` 검증, `env.json` 기록.

---

## 실험·검증(Validation) 기준
| 항목 | 측정/검증 방법 | 통과 기준 |
| --- | --- | --- |
| **학습 성공률** | 01/02 노트북 | 중단/NaN 0건 |
| **지표 수렴** | `train_log.csv` | 5 epoch 이후 val Macro-F1 상승 추세 |
| **베스트 선정** | `macro_f1_curve.png` | 과도한 초반/말단 치우침 없음 |
| **KFold 일관성** | fold별 요약 | Macro-F1 표준편차 ≤ **±2.0%p** |
| **재현성** | 동일 설정 2회 | 최종 Macro-F1 편차 ≤ **±1.0%p** |
| **Resume 일관성** | 3ep 중단→재개 | 최종 Macro-F1 **±0.1%p** |
| **체크포인트 스키마** | 필수 키 검사 | 누락 0건 |
| **해시 잠금** | sha256 검증 | 실행 전후 동일 |
| **아티팩트 완결성** | 파일 존재 검사 | config/env/summary/log/plots/ckpt 모두 존재 |

---

## 병합 조건(Definition of Done & PR 체크리스트)
- **DoD**
    - ✅ 00~07 노트북 **오류 0, 출력 포함**
    - ✅ `experiments/train/<run_id>/…` 아티팩트 전부 생성
    - ✅ 단일/다중 폴드에서 Macro-F1 기준 수렴
    - ✅ **Resume 일관성**·**해시 잠금** 기준 충족
    - ✅ 규칙 준수(외부 데이터/PL/유료 API 미사용)
- **PR 체크리스트**
    - [ ]  black/isort, 린트 통과
    - [ ]  단위 테스트 5종 통과
        - `test_model_output_shape.py`
        - `test_minibatch_train.py`
        - `test_checkpoint_schema.py`
        - `test_resume.py`
        - `test_param_groups.py`
    - [ ]  `configs/train_config.json` 커밋(+ `config.hash`/`folds.hash` 첨부)
    - [ ]  아티팩트·로그 링크(요약 스크린샷 포함)

---

## 브랜치 작업 순서(권장 커밋 플로우)
1. `feat(config)`: `train_config.json`/런 아이디/경로·해시 잠금
2. `feat(backbone)`: timm 백본 빌더 + 헤드
3. `feat(optim-sched)`: AdamW Param-Group + 스케줄러
4. `feat(loop)`: AMP/클리핑/EMA/Failover/metrics hook
5. `feat(callbacks)`: 체크포인트/ES/HistoryLogger
6. `feat(single-fold)`: 단일 폴드 학습(01)
7. `feat(kfold)`: 폴드 러너 + (옵션) OOF 훅(02)
8. `feat/ablation`(03) → `feat(plots)`(04)
9. `feat(export)`(05)
10. `feat(resume)`(06) · `feat(cli-ci)`(07)
11. `tests/*` 확장

---

## 구성·설정 표준 (`configs/train_config.json` 예시)
```json
{
  "arch": "resnet50",
  "img_size": 224,
  "num_classes": 17,
  "epochs": 20,
  "batch_size": 64,
  "optimizer": {
    "name": "adamw",
    "lr": 0.0003,
    "weight_decay": 0.0001,
    "no_decay_norm_bias": true
  },
  "scheduler": { "name": "cosine", "warmup_pct": 0.05, "t_max": "epochs" },
  "loss": {
    "name": "ce",
    "label_smoothing": 0.1,
    "use_class_weights": true,
    "focal_gamma": null
  },
  "sampler": { "use_weighted": true },
  "regularization": {
    "grad_clip_norm": 1.0,
    "dropout": 0.2,
    "ema": true,
    "ema_decay": 0.999,
    "ema_on_eval": true
  },
  "amp": true,
  "precision": "fp16",
  "accumulation_steps": 1,
  "eval_interval": 1,
  "save_every_n_epochs": 1,
  "resume_from": null,
  "strict_resume": true,
  "max_loader_errors": 0,
  "paths": {
    "train_csv": "data/raw/train.csv",
    "img_root": "data/raw/train",
    "folds_csv": "data/processed/folds_k5_seed42.csv",
    "class_weights": "data/processed/class_weights.json",
    "exp_dir": "experiments/train"
  },
  "seed": 42
}
```

---

## 파라미터 표준(요약)
| 항목 | 기본값 | 권장 범위 | 비고 |
| --- | --- | --- | --- |
| lr | 3e-4 | 1e-4 ~ 1e-3 | AdamW |
| weight_decay | 1e-4 | 0 ~ 5e-4 | bias/Norm no_decay |
| epochs | 20 | 15 ~ 50 | Cosine+Warmup |
| batch_size | 64 | 32 ~ 128 | AMP/Accum 병행 |
| label_smoothing | 0.1 | 0.05 ~ 0.2 | CE 계열 |
| focal_gamma | – | 1 ~ 2 | 선택 |
| grad_clip_norm | 1.0 | 0.5 ~ 1.0 | global-norm |
| ema_decay | 0.999 | 0.995 ~ 0.9995 | ep≥3부터 |
| precision | fp16 | bf16/fp32 | 집계는 fp32 |

---

## 리스크·준수사항
- **규칙 준수:** 외부 데이터/PL/유료 API 미사용, **테스트 데이터 학습·튜닝 금지**.
- **재현성 vs 성능:** `cudnn.benchmark` ON은 속도↑·결정성↓ → 기본 OFF, 필요 시 로그에 명시.
- **리소스 제약:** 고해상도/대배치 OOM 위험 → AMP/Accum·배치/해상도 조절.
- **체크포인트 부피:** top-k 과다 저장 시 디스크 압박 → `save_every_n_epochs`로 제어.
- **OOF 훅 사용 시:** 검증셋·OOF만 저장, **테스트 데이터는 일절 포함 금지**.
